operator/pkg/awsprovider/iam/reconciler.go

```diff
@@ -0,0 +1,210 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iam
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/iam"
+	apis "github.com/awslabs/kit/operator/pkg/apis/controlplane/v1alpha1"
+	"github.com/awslabs/kit/operator/pkg/awsprovider"
+	"github.com/awslabs/kit/operator/pkg/errors"
+	"github.com/awslabs/kit/operator/pkg/kubeprovider"
+	"go.uber.org/zap"
+
+	"knative.dev/pkg/ptr"
+)
+
+type Controller struct {
+	iam        *awsprovider.IAM
+	kubeClient *kubeprovider.Client
+}
+
+// NewController returns a controller for managing LaunchTemplates and ASG in AWS
+func NewController(iam *awsprovider.IAM, client *kubeprovider.Client) *Controller {
+	return &Controller{iam: iam, kubeClient: client}
+}
+
+var (
+	desiredPolicies = []string{
+		"arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy",
+		"arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
+		"arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy",
+		"arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly",
+	}
+)
+
+func (c *Controller) Reconcile(ctx context.Context, controlPlane *apis.ControlPlane) error {
+	role, err := c.getRole(ctx, controlPlane)
+	if err != nil && !errors.IsIAMObjectDoNotExist(err) {
+		return fmt.Errorf("getting IAM role for %v, %w", controlPlane.ClusterName(), err)
+	}
+	if role == nil {
+		role, err = c.createRole(ctx, controlPlane)
+		if err != nil {
+			return fmt.Errorf("creating IAM role for %v, %w", controlPlane.ClusterName(), err)
+		}
+		zap.S().Infof("[%s] Created IAM Role %v", controlPlane.ClusterName(), aws.StringValue(role.RoleName))
+	}
+	if err := role.addRoleToInstanceProfile(ctx, controlPlane); err != nil {
+		return fmt.Errorf("adding instance profile to role, %w", err)
+	}
+	for _, policy := range desiredPolicies {
+		if err := role.attachPolicy(ctx, policy, controlPlane.ClusterName()); err != nil {
+			return fmt.Errorf("attaching policies to role, %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) Finalize(ctx context.Context, controlPlane *apis.ControlPlane) error {
+	_, err := c.iam.RemoveRoleFromInstanceProfileWithContext(ctx, &iam.RemoveRoleFromInstanceProfileInput{
+		InstanceProfileName: aws.String(KitNodeInstanceProfileNameFor(controlPlane.ClusterName())),
+		RoleName:            aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
```

Comment:
function name sounds generic but this function is only good for `kitNodeROle` based on this line. Instead of passing `controlPlane *apis.ControlPlane` as func param, i would recommend passing `rolename` as arg to this function. Thoughts ?

Reply:
not clear which function you are refering to here `KitNodeRoleNameFor`?